presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -552,6 +558,13 @@ public boolean transitionToFinishing()
             return false;
         }
 
+        try {
+            metadata.endQuery(session);
```

Comment:
Shouldn't `endQuery` be triggered after `commit` (if auto-commit is on), @electrum?